vortex-array/src/operator/mod.rs

```diff
@@ -0,0 +1,200 @@
+/// deep equality or hashing would be too expensive.
+pub trait Operator: 'static + Send + Sync + Debug + DynEq + DynHash {
+    /// The unique identifier for this operator instance.
+    fn id(&self) -> OperatorId;
+
+    /// For downcasting.
+    fn as_any(&self) -> &dyn Any;
+
+    /// Returns the [`DType`] of the array produced by this operator.
+    fn dtype(&self) -> &DType;
+
+    /// Returns the number of rows produced by this operator.
+    fn len(&self) -> usize;
+
+    /// Returns whether this operator produces zero rows.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    // TODO(ngates): add StatsSet?
+
+    /// The children of this operator.
+    fn children(&self) -> &[OperatorRef];
+
+    /// The number of children of this operator.
+    fn nchildren(&self) -> usize {
+        self.children().len()
+    }
+
+    /// Override the default formatting of this operator.
+    fn fmt_as(&self, _df: DisplayFormat, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", type_name::<Self>())
+    }
+
+    /// Create a new instance of this operator with the given children.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if the number or dtypes of children are incorrect.
+    ///
+    fn with_children(self: Arc<Self>, _children: Vec<OperatorRef>) -> VortexResult<OperatorRef>;
+
+    /// Attempt to optimize this node by analyzing its children.
+    ///
+    /// For example, if all the children are constant, this function should perform constant
+    /// folding and return a constant operator.
+    ///
+    /// This function should typically be implemented only for self-contained optimizations based
+    /// on child properties
+    fn reduce_children(&self) -> VortexResult<Option<OperatorRef>> {
+        Ok(None)
+    }
+
+    /// Attempt to push down a parent operator through this node.
+    ///
+    /// The `child_idx` parameter indicates which child of the parent this operator occupies.
+    /// For example, if the parent is a binary operator, and this operator is the left child,
+    /// then `child_idx` will be 0. If this operator is the right child, then `child_idx` will be 1.
+    ///
+    /// The returned operator will replace the parent in the tree.
+    ///
+    /// This function should typically be implemented for cross-operator optimizations where the
+    /// child needs to adapt to the parent's requirements
+    fn reduce_parent(
+        &self,
+        _parent: OperatorRef,
+        _child_idx: usize,
+    ) -> VortexResult<Option<OperatorRef>> {
+        Ok(None)
+    }
+
+    // Which fields this operator accesses (for projection pushdown)
+    // fn required_columns(&self) -> Option<Vec<FieldPath>> {
+    //     None
+    // }
+
+    /// Whether this operator is aligned 1:1 with its child.
+    ///
+    /// Returns `None` if unknown.
+    fn is_position_preserving(&self, _child_idx: usize) -> Option<bool> {
+        None
+    }
```

Comment:
I think we might end up with lots of these shall now put them in another place or add todo?